dwave/cloud/solver.py

```diff
@@ -200,6 +223,166 @@ def is_online(self):
         warnings.warn("'is_online' property is deprecated in favor of 'online'.", DeprecationWarning)
         return self.online
 
+
+class UnstructuredSolver(BaseSolver):
+    """Class for D-Wave unstructured solvers.
+
+    This class provides :term:`Ising`, :term:`QUBO` and :term:`BQM` sampling
+    methods and encapsulates the solver description returned from the D-Wave
+    cloud API.
+
+    Args:
+        client (:class:`~dwave.cloud.client.Client`):
+            Client that manages access to this solver.
+
+        data (`dict`):
+            Data from the server describing this solver.
+
+    """
+
+    _handled_problem_types = {"bqm"}
+    _handled_encoding_formats = {"bq"}
+
+    def sample_ising(self, linear, quadratic, **params):
+        """Sample from the specified :term:`BQM`.
+
+        Args:
+            bqm (:class:`~dimod.BinaryQuadraticModel`):
+                A binary quadratic model.
+
+            **params:
+                Parameters for the sampling method, solver-specific.
+
+        Returns:
+            :class:`Future`
+
+        Note:
+            To use this method, dimod package has to be installed.
+        """
+        try:
+            import dimod
+        except ImportError: # pragma: no cover
+            raise RuntimeError("Can't use solver of type 'bqm' without dimod. "
+                               "Re-install the library with 'bqm' support.")
+
+        bqm = dimod.BinaryQuadraticModel.from_ising(linear, quadratic)
+        return self.sample_bqm(bqm, **params)
+
+    def sample_qubo(self, qubo, **params):
+        """Sample from the specified :term:`QUBO`.
+
+        Args:
+            qubo (dict[(int, int), float]):
+                Coefficients of a quadratic unconstrained binary optimization
+                (QUBO) model.
+
+            **params:
+                Parameters for the sampling method, solver-specific.
+
+        Returns:
+            :class:`Future`
+
+        Note:
+            To use this method, dimod package has to be installed.
+        """
+        try:
+            import dimod
+        except ImportError: # pragma: no cover
+            raise RuntimeError("Can't use solver of type 'bqm' without dimod. "
+                               "Re-install the library with 'bqm' support.")
+
+        bqm = dimod.BinaryQuadraticModel.from_qubo(qubo)
+        return self.sample_bqm(bqm, **params)
+
+    def sample_bqm(self, bqm, **params):
```

Comment:
Should you also raise a RuntimeError like the other sample methods explaining this needs dimod? Maybe it's not necessary because it explains what the `bqm` param is supposed to be in the docstring...

Reply:
Strictly speaking, for this method, dimod is not required. That is, dimod calls are not used anywhere on the "submit-this-bqm-to-sapi" code path. The only requirement is for the input `bqm` object to have `.to_serializeable()`.

That's unlike other sample methods here, in which we use dimod to construct a BQM object.